### PR TITLE
fix(runtime/storage): root mismatch error

### PR DESCRIPTION
**Pull request by jimjbrettj**
## Changes

- When calling NextKey, if we are in the middle of a transaction the function returns the value, rather than the key itself, which in turn causes an invalid state and causes us to get a root mismatch error. This PR changes this behavior by returning the key instead of the value
- Similar errors occurred in the functions GetKeysWithPrefixFromChild and GetChildNextKey, and these have been modified to the correct behavior as well
- Wrote a test testing the block execution at the specific block where the problem occurred
- In the storage tests, we used the same values for both the keys, values, and keys to child in many tests, which is how this bug did not get detected until now. The test has been updated to use separate values

## Tests

```sh
go test ./lib/runtime/storage -v
```

## Issues

#3943 

## Primary Reviewer

@dimartiro

### CODE REVIEW

File: lib/runtime/storage/trie.go [hunk 1]
```diff
@@ -466,7 +470,8 @@ func (t *TrieState) GetChildNextKey(keyToChild, key []byte) ([]byte, error) {
 
 			for _, k := range keys {
 				if k > string(key) && !childChanges.deletes[k] {
-					return allEntries[k], nil
+					return []byte(k), nil
+					//return allEntries[k], nil
```
Comment (EclesioMeloJunior) [inline lib/runtime/storage/trie.go]: Remove

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM! Also nice you improved the tests!

File: lib/runtime/storage/trie.go [hunk 2]
```diff
@@ -146,6 +146,9 @@ func (t *TrieState) MustRoot() common.Hash {
 // Root returns the trie's root hash
 func (t *TrieState) Root() (common.Hash, error) {
 	// Since the Root function is called without running transactions we can do:
+	if currentTx := t.getCurrentTransaction(); currentTx != nil {
+		panic("cannot calculate root with running transactions")
+	}
```
Comment (dimartiro) [inline lib/runtime/storage/trie.go:151]: This should never happen, but it's really nice to add the check! thanks!

Comment (dimartiro) [inline lib/runtime/storage/trie.go:151]: This should never happen, but it's really nice to add the check! thanks!

File: lib/runtime/storage/trie.go [hunk 3]
```diff
@@ -181,7 +184,7 @@ func (t *TrieState) NextKey(key []byte) []byte {
 
 		for _, k := range keys {
 			if k > string(key) && !currentTx.deletes[k] {
-				return allEntries[k]
+				return []byte(k)
```
Comment (dimartiro) [inline lib/runtime/storage/trie.go:187]: my bad 😅

**Review by dimartiro — APPROVED** [review]
Looks good! thanks for adding more tests for this case an sorry for the issues 😅

**Pull request merged**